README
NAME

Lexical::Var - static variables without namespace pollution

DESCRIPTION

This module implements lexical scoping of static variables and
subroutines.  Although it can be used directly, it is mainly intended
to be infrastructure for modules that manage namespaces.

This module influences the meaning of single-part variable names that
appear directly in code, such as "$foo".  Normally, in the absence of any
particular declaration, or under the effect of an "our" declaration, this
would refer to the scalar variable of that name located in the current
package.  A "Lexical::Var" declaration can change this to refer to any
particular scalar, bypassing the package system entirely.  A variable
name that includes an explicit package part, such as "$main::foo", always
refers to the variable in the specified package, and is unaffected by this
module.  A symbolic reference through a string value, such as "${'foo'}",
also looks in the package system, and so is unaffected by this module.

The types of name that can be influenced are scalar ("$foo"), array
("@foo"), hash ("%foo"), subroutine ("&foo"), and glob ("*foo").
A definition for any of these names also affects code that logically
refers to the same entity, even when the name is spelled without its
usual sigil.  For example, any definition of "@foo" affects element
references such as "$foo[0]".  Barewords in filehandle context actually
refer to the glob variable.  Bareword references to subroutines, such as
"foo(123)", only work on Perl 5.11.2 and later; on earlier Perls you
must use the "&" sigil, as in "&foo(123)".

Where a scalar name is defined to refer to a constant (read-only) scalar,
references to the constant through the lexical namespace can participate
in compile-time constant folding.  This can avoid the need to check
configuration values (such as whether debugging is enabled) at runtime.

A name definition supplied by this module takes effect from the end
of the definition statement up to the end of the immediately enclosing
block, except where it is shadowed within a nested block.  This is the
same lexical scoping that the "my", "our", and "state" keywords supply.
Definitions from Lexical::Var and from "my"/"our"/"state" can shadow
each other.  These lexical definitions propagate into string "eval"s,
on Perl versions that support it (5.9.3 and later).

This module only manages variables of static duration (the kind of
duration that "our" and "state" variables have).  To get a fresh variable
for each invocation of a function, use "my".

INSTALLATION

	perl Build.PL
	./Build
	./Build test
	./Build install

AUTHOR

Andrew Main (Zefram) <zefram@fysh.org>

COPYRIGHT

Copyright (C) 2009, 2010, 2011, 2012
Andrew Main (Zefram) <zefram@fysh.org>

LICENSE

This module is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


Changes
version 0.008; 2013-08-17

  * update for perl 5.17.4's new pad API

  * update for perl 5.17.5's new arrangement for PL_compcv with BEGIN

  * in documentation, refer to the new module Scalar::Construct

version 0.007; 2012-02-04

  * bugfix: preserve referential identity of named constant scalars on
    threading Perl builds

  * bugfix: require bugfixed version of Lexical::SealRequireHints (for
    compatibility with early-loaded warnings.pm) and invoke it earlier
    to make sure it takes effect in time

  * in documentation, note the problems that threading Perls have with
    objects originating from source literals

  * in documentation, note that the problem with eval/require/do inside
    a BEGIN block is fixed in Perl 5.15.5

  * in documentation, note that constant scalars participate in constant
    folding

  * fix an outdated documentation note about bareword subroutine calls

  * in documentation, tweak description of intended users

  * additional tests regarding const ops, constant folding, and object
    identity preservation

  * additional tests for the Lexical::Sub interface

  * in Build.PL, declare incompatibility with pre-0.19
    B::Hooks::OP::Check, which doesn't play nicely around op check hooking

  * reorganise some of the test suite

  * convert .cvsignore to .gitignore

  * sort MANIFEST

version 0.006; 2011-07-27

  * bugfix: require bugfixed version of Lexical::SealRequireHints (for
    working around [perl #73174])

  * in documentation, note problem with indirect object syntax and
    lexical subroutines

  * include META.json in distribution

  * add MYMETA.json to .cvsignore

version 0.005; 2011-02-27

  * port to Perl 5.13.10, where the value of PAD_MAX has changed

  * in XS, use gv_stashpvs() wherever appropriate

  * in XS, use PERL_NO_GET_CONTEXT for efficiency

  * in XS, declare "PROTOTYPES: DISABLE" to prevent automatic generation
    of unintended prototypes

  * jump through hoops to avoid compiler warnings

  * use full stricture in test suite

  * in test suite, make all numeric comparisons against $] stringify it
    first, to avoid architecture-dependent problems with floating point
    rounding giving it an unexpected numeric value

  * in Build.PL, complete declaration of configure-time requirements

version 0.004; 2010-04-11

  * bugfix: require bugfixed version of Lexical::SealRequireHints (for
    passing package through to required code in pure-Perl version of
    Lexical::SealRequireHints)

  * in XS, use macros to avoid explicit passing of aTHX, in the manner
    of the core

  * add Lexical::Import to "see also" list

  * add MYMETA.yml to .cvsignore

version 0.003; 2010-01-11

  * fully support bareword subroutine references on Perl 5.11.2 and later

  * in XS, avoid using "class" as a variable name, for compatibility
    with C++ compilers

  * correct data type for the PAD_MAX constant

  * in Build.PL, explicitly declare configure-time requirements

version 0.002; 2009-10-26

  * port to Perl 5.11.0, supporting the addition of first-class regexp
    objects (which are actually a type of scalar) and the removal of
    the distinct RV type

  * where the compilation %^H and PL_compcv are unavailable due to string
    eval or similar, detect this earlier and regardless of the type of
    item being set up (previously it was not detected for subroutines
    or globs)

  * when a lexical scalar has readonly value, generate a const op instead
    of the usual rv2sv

  * document the problem of %^H and PL_compcv being unavailable due to
    string eval

  * preserve op flags when building new rv2Xv ops; this doesn't seem to
    matter in any currently achievable situation, but will be required
    for bareword subroutine references in the future

  * in XS, properly parenthesise some macro definitions

  * test writability of lexical variables

  * in t/error.t, check for unexpected warnings

  * add experimental code, all disabled, to support bareword subroutine
    references on a suitably modified Perl core

version 0.001; 2009-09-30

  * bugfix: make array and hash indexing work

  * bugfix: avoid core checks that would reject use of a glob as a
    scalar value

  * bugfix: avoid modifying the reference stored in the hint hash (which
    happened through reusing it in const ops)

  * bugfix: repair a reference leak in unimportation

  * bugfix: change a memNE() to strnNE(), to avoid reading off the end
    of a string

  * bugfix: require bugfixed version of Lexical::SealRequireHints (for
    fixes around usability of its pure-Perl implementation)

  * port to pre-5.10 Perls

  * more tests: Lexical::Sub, array and hash variables, glob variables,
    acceptability of value types for references to lexical variables,
    erroneous invocation of import and unimport

  * in XS, add some casts to avoid compiler warnings

  * check for required Perl version at runtime

version 0.000; 2009-09-23

  * initial released version


t/scalar_1.pm
use strict;
use Lexical::Var '$foo' => \2;
1;


t/scalar_2.pm
use strict;
use Lexical::Var '$foo' => \2;
push @main::values, $foo;
1;


Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.3601
require 5.006;

    unless (eval "use Module::Build::Compat 0.02; 1" ) {
      print "This module requires Module::Build to install itself.\n";

      require ExtUtils::MakeMaker;
      my $yn = ExtUtils::MakeMaker::prompt
	('  Install Module::Build now from CPAN?', 'y');

      unless ($yn =~ /^y/i) {
	die " *** Cannot install without Module::Build.  Exiting ...\n";
      }

      require Cwd;
      require File::Spec;
      require CPAN;

      # Save this 'cause CPAN will chdir all over the place.
      my $cwd = Cwd::cwd();

      CPAN::Shell->install('Module::Build::Compat');
      CPAN::Shell->expand("Module", "Module::Build::Compat")->uptodate
	or die "Couldn't install Module::Build, giving up.\n";

      chdir $cwd or die "Cannot chdir() back to $cwd: $!";
    }
    eval "use Module::Build::Compat 0.02; 1" or die $@;
    
    Module::Build::Compat->run_build_pl(args => \@ARGV);
    my $build_script = 'Build';
    $build_script .= '.com' if $^O eq 'VMS';
    exit(0) unless(-e $build_script); # cpantesters convention
    require Module::Build;
    Module::Build::Compat->write_makefile(build_class => 'Module::Build');


lib/Lexical/Sub.pm
=head1 NAME

Lexical::Sub - subroutines without namespace pollution

=head1 SYNOPSIS

	use Lexical::Sub quux => sub { $_[0] + 1 };
	use Lexical::Sub carp => \&Carp::carp;

=head1 DESCRIPTION

This module implements lexical scoping of subroutines.  Although it can
be used directly, it is mainly intended to be infrastructure for modules
that manage namespaces.

This module influences the meaning of single-part subroutine names that
appear directly in code, such as "C<&foo>" and "C<foo(123)>".
Normally, in the absence of
any particular declaration, these would refer to the subroutine of that
name located in the current package.  A C<Lexical::Sub> declaration
can change this to refer to any particular subroutine, bypassing the
package system entirely.  A subroutine name that includes an explicit
package part, such as "C<&main::foo>", always refers to the subroutine
in the specified package, and is unaffected by this module.  A symbolic
reference through a string value, such as "C<&{'foo'}>", also looks in
the package system, and so is unaffected by this module.

Bareword references to subroutines, such as "C<foo(123)>", only work on
Perl 5.11.2 and later.  On earlier Perls you must use the C<&> sigil,
as in "C<&foo(123)>".

A name definition supplied by this module takes effect from the end of the
definition statement up to the end of the immediately enclosing block,
except where it is shadowed within a nested block.  This is the same
lexical scoping that the C<my>, C<our>, and C<state> keywords supply.
These lexical definitions propagate into string C<eval>s, on Perl versions
that support it (5.9.3 and later).

This module is implemented through the mechanism of L<Lexical::Var>.
Its distinct name and declaration syntax exist to make lexical subroutine
declarations clearer.

=cut

package Lexical::Sub;

{ use 5.006; }
use warnings;
use strict;

our $VERSION = "0.008";

require Lexical::Var;
die "mismatched versions of Lexical::Var and Lexical::Sub modules"
	unless $Lexical::Var::VERSION eq $VERSION;

=head1 PACKAGE METHODS

These methods are meant to be invoked on the C<Lexical::Sub> package.

=over

=item Lexical::Sub->import(NAME => REF, ...)

Sets up lexical subroutine declarations, in the lexical environment that
is currently compiling.  Each I<NAME> must be a bare subroutine name
(e.g., "B<foo>"), and each I<REF> must be a reference to a subroutine.
The name is lexically associated with the referenced subroutine.

=item Lexical::Sub->unimport(NAME [=> REF], ...)

Sets up negative lexical subroutine declarations, in the lexical
environment that is currently compiling.  Each I<NAME> must be a bare
subroutine name (e.g., "B<foo>").  If the name is given on its own, it is
lexically dissociated from any subroutine.  Within the resulting scope,
the subroutine name will not be recognised.  If a I<REF> (which must
be a reference to a subroutine) is specified with a name, the name
will be dissociated if and only if it is currently associated with
that subroutine.

=back

=head1 BUGS

Subroutine invocations without the C<&> sigil cannot be correctly
processed on Perl versions earlier than 5.11.2.  This is because
the parser needs to look up the subroutine early, in order to let any
prototype affect parsing, and it looks up the subroutine by a different
mechanism than is used to generate the call op.  (Some forms of sigilless
call have other complications of a similar nature.)  If an attempt
is made to call a lexical subroutine via a bareword on an older Perl,
this module will probably still be able to intercept the call op, and
will throw an exception to indicate that the parsing has gone wrong.
However, in some cases compilation goes further wrong before this
module can catch it, resulting in either a confusing parse error or
(in rare situations) silent compilation to an incorrect op sequence.
On Perl 5.11.2 and later, sigilless subroutine calls work correctly,
except for an issue noted below.

Subroutine calls that have neither sigil nor parentheses (around the
argument list) are subject to an ambiguity with indirect object syntax.
If the first argument expression begins with a bareword or a scalar
variable reference then the Perl parser is liable to interpret the call as
an indirect method call.  Normally this syntax would be interpreted as a
subroutine call if the subroutine exists, but the parser doesn't look at
lexically-defined subroutines for this purpose.  The call interpretation
can be forced by prefixing the first argument expression with a C<+>,
or by wrapping the whole argument list in parentheses.

Package hash entries get created for subroutine names that are used,
even though the subroutines are not actually being stored or looked
up in the package.  This can occasionally result in a "used only once"
warning failing to occur when it should.

On Perls prior to 5.15.5,
if this package's C<import> or C<unimport> method is called from inside 
a string C<eval> inside a C<BEGIN> block, it does not have proper
access to the compiling environment, and will complain that it is being
invoked outside compilation.  Calling from the body of a C<require>d
or C<do>ed file causes the same problem
on the same Perl versions.  Other kinds of indirection
within a C<BEGIN> block, such as calling via a normal function, do not
cause this problem.

=head1 SEE ALSO

L<Lexical::Import>,
L<Lexical::Var>

=head1 AUTHOR

Andrew Main (Zefram) <zefram@fysh.org>

=head1 COPYRIGHT

Copyright (C) 2009, 2010, 2011, 2012, 2013
Andrew Main (Zefram) <zefram@fysh.org>

=head1 LICENSE

This module is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

1;


lib/Lexical/Var.pm
=head1 NAME

Lexical::Var - static variables without namespace pollution

=head1 SYNOPSIS

	use Lexical::Var '$foo' => \$Remote::foo;
	use Lexical::Var '$const' => \123;
	use Lexical::Var '@bar' => [];
	use Lexical::Var '%baz' => { a => 1, b => 2 };
	use Lexical::Var '&quux' => sub { $_[0] + 1 };
	use Lexical::Var '*wibble' => Symbol::gensym();

=head1 DESCRIPTION

This module implements lexical scoping of static variables and
subroutines.  Although it can be used directly, it is mainly intended
to be infrastructure for modules that manage namespaces.

This module influences the meaning of single-part variable names that
appear directly in code, such as "C<$foo>".  Normally, in the absence
of any particular declaration, or under the effect of an C<our>
declaration, this would refer to the scalar variable of that name
located in the current package.  A C<Lexical::Var> declaration can
change this to refer to any particular scalar, bypassing the package
system entirely.  A variable name that includes an explicit package part,
such as "C<$main::foo>", always refers to the variable in the specified
package, and is unaffected by this module.  A symbolic reference through
a string value, such as "C<${'foo'}>", also looks in the package system,
and so is unaffected by this module.

The types of name that can be influenced are scalar ("C<$foo>"),
array ("C<@foo>"), hash ("C<%foo>"), subroutine ("C<&foo>"), and glob
("C<*foo>").  A definition for any of these names also affects code
that logically refers to the same entity, even when the name is spelled
without its usual sigil.  For example, any definition of "C<@foo>" affects
element references such as "C<$foo[0]>".  Barewords in filehandle context
actually refer to the glob variable.  Bareword references to subroutines,
such as "C<foo(123)>", only work on Perl 5.11.2 and later; on earlier
Perls you must use the C<&> sigil, as in "C<&foo(123)>".

Where a scalar name is defined to refer to a constant (read-only) scalar,
references to the constant through the lexical namespace can participate
in compile-time constant folding.  This can avoid the need to check
configuration values (such as whether debugging is enabled) at runtime.

A name definition supplied by this module takes effect from the end of the
definition statement up to the end of the immediately enclosing block,
except where it is shadowed within a nested block.  This is the same
lexical scoping that the C<my>, C<our>, and C<state> keywords supply.
Definitions from L<Lexical::Var> and from C<my>/C<our>/C<state> can shadow
each other.  These lexical definitions propagate into string C<eval>s,
on Perl versions that support it (5.9.3 and later).

This module only manages variables of static duration (the kind of
duration that C<our> and C<state> variables have).  To get a fresh
variable for each invocation of a function, use C<my>.

=cut

package Lexical::Var;

{ use 5.006; }
use Lexical::SealRequireHints 0.006;
use warnings;
use strict;

our $VERSION = "0.008";

require XSLoader;
XSLoader::load(__PACKAGE__, $VERSION);

=head1 PACKAGE METHODS

These methods are meant to be invoked on the C<Lexical::Var> package.

=over

=item Lexical::Var->import(NAME => REF, ...)

Sets up lexical variable declarations, in the lexical environment that
is currently compiling.  Each I<NAME> must be a variable name (e.g.,
"B<$foo>") including sigil, and each I<REF> must be a reference to a
variable/value of the appropriate type.  The name is lexically associated
with the referenced variable/value.

L<Scalar::Construct> can be helpful in generating appropriate I<REF>s,
especially to create constants.  There are Perl core bugs to beware of
around compile-time constants; see L</BUGS>.

=item Lexical::Var->unimport(NAME [=> REF], ...)

Sets up negative lexical variable declarations, in the lexical environment
that is currently compiling.  Each I<NAME> must be a variable name
(e.g., "B<$foo>") including sigil.  If the name is given on its own,
it is lexically dissociated from any value.  Within the resulting scope,
the variable name will not be recognised.  If a I<REF> (which must be a
reference to a value of the appropriate type) is specified with a name,
the name will be dissociated if and only if it is currently associated
with that value.

=back

=head1 BUGS

Subroutine invocations without the C<&> sigil cannot be correctly
processed on Perl versions earlier than 5.11.2.  This is because
the parser needs to look up the subroutine early, in order to let any
prototype affect parsing, and it looks up the subroutine by a different
mechanism than is used to generate the call op.  (Some forms of sigilless
call have other complications of a similar nature.)  If an attempt
is made to call a lexical subroutine via a bareword on an older Perl,
this module will probably still be able to intercept the call op, and
will throw an exception to indicate that the parsing has gone wrong.
However, in some cases compilation goes further wrong before this
module can catch it, resulting in either a confusing parse error or
(in rare situations) silent compilation to an incorrect op sequence.
On Perl 5.11.2 and later, sigilless subroutine calls work correctly,
except for an issue noted below.

Subroutine calls that have neither sigil nor parentheses (around the
argument list) are subject to an ambiguity with indirect object syntax.
If the first argument expression begins with a bareword or a scalar
variable reference then the Perl parser is liable to interpret the call as
an indirect method call.  Normally this syntax would be interpreted as a
subroutine call if the subroutine exists, but the parser doesn't look at
lexically-defined subroutines for this purpose.  The call interpretation
can be forced by prefixing the first argument expression with a C<+>,
or by wrapping the whole argument list in parentheses.

On Perls built for threading (even if threading is not actually used),
scalar constants that are defined by literals in the Perl source don't
reliably maintain their object identity.  What appear to be multiple
references to a single object can end up behaving as references
to multiple objects, in surprising ways.  The multiple objects all
initially have the correct value, but they can be writable even though the
original object is a constant.  See Perl bug reports [perl #109744] and
[perl #109746].  This can affect objects that are placed in the lexical
namespace, just as it can affect those in package namespaces or elsewhere.
C<Lexical::Var> avoids contributing to the problem itself, but certain
ways of building the parameters to C<Lexical::Var> can result in the
object in the lexical namespace not being the one that was intended,
or can damage the named object so that later referencing operations on
it misbehave.  L<Scalar::Construct> can be used to avoid this problem.

Bogus redefinition warnings occur in some cases when C<our> declarations
and C<Lexical::Var> declarations shadow each other.

Package hash entries get created for subroutine and glob names that
are used, even though the subroutines and globs are not actually being
stored or looked up in the package.  This can occasionally result in a
"used only once" warning failing to occur when it should.

On Perls prior to 5.15.5,
if this package's C<import> or C<unimport> method is called from inside
a string C<eval> inside a C<BEGIN> block, it does not have proper
access to the compiling environment, and will complain that it is being
invoked outside compilation.  Calling from the body of a C<require>d
or C<do>ed file causes the same problem
on the same Perl versions.  Other kinds of indirection
within a C<BEGIN> block, such as calling via a normal function, do not
cause this problem.

=head1 SEE ALSO

L<Attribute::Lexical>,
L<Lexical::Import>,
L<Lexical::Sub>,
L<Scalar::Construct>

=head1 AUTHOR

Andrew Main (Zefram) <zefram@fysh.org>

=head1 COPYRIGHT

Copyright (C) 2009, 2010, 2011, 2012, 2013
Andrew Main (Zefram) <zefram@fysh.org>

=head1 LICENSE

This module is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

1;
